servicetalk-grpc-netty/src/test/java/io/servicetalk/grpc/netty/GrpcMessageEncodingTest.java

```diff
@@ -0,0 +1,410 @@
+                {null, null, GZIP, false},
+                {null, null, DEFLATE, false},
+                {of(NONE, GZIP, DEFLATE), null, NONE, true},
+                {of(NONE, GZIP, DEFLATE), null, GZIP, true},
+                {of(NONE, GZIP, DEFLATE), null, DEFLATE, true},
+                {of(NONE, GZIP), null, DEFLATE, false},
+                {of(NONE, DEFLATE), null, GZIP, false},
+                {of(NONE, DEFLATE), null, DEFLATE, true},
+                {of(NONE, DEFLATE), null, NONE, true},
+                {of(GZIP), null, NONE, true},
+                {of(GZIP), of(GZIP), NONE, true},
+                {of(GZIP), of(NONE, GZIP), NONE, true},
+                {of(GZIP), of(NONE, GZIP), GZIP, true},
+                {null, of(NONE, GZIP), GZIP, false},
+                {null, of(NONE, GZIP), DEFLATE, false},
+                {null, of(NONE, GZIP), NONE, true},
+        };
+    }
+
+    @After
+    public void tearDown() throws Exception {
+        try {
+            client.close();
+        } finally {
+            serverContext.close();
+        }
+    }
+
+    private ServerContext listenAndAwait(final TestEncodingScenario encodingOptions) throws Exception {
+
+        StreamingHttpServiceFilterFactory filterFactory = REQ_RESP_VERIFIER.apply(encodingOptions);
+        if (encodingOptions.serverSupported == null) {
+            return grpcServerBuilder.appendHttpServiceFilter(filterFactory)
+                                             .listenAndAwait(new ServiceFactory(new TesterServiceImpl()));
+        } else {
+            return grpcServerBuilder.appendHttpServiceFilter(filterFactory)
+                                             .listenAndAwait(new ServiceFactory(new TesterServiceImpl(),
+                                                     encodingOptions.serverSupported));
+        }
+    }
+
+    private TesterClient newClient(@Nullable final Set<GrpcMessageEncoding> supportedEncodings) {
+        return GrpcClients.forAddress(serverHostAndPort(serverContext))
+                .executionStrategy(noOffloadsStrategy())
+                .build(supportedEncodings != null ?
+                        new ClientFactory().supportedEncodings(supportedEncodings) :
+                        new ClientFactory());
+    }
+
+    @Test
+    public void test() throws ExecutionException, InterruptedException {
+        if (expectedSuccess) {
+            assertSuccessful(requestEncoding);
+        } else {
+            assertUnimplemented(requestEncoding);
+        }
+    }
+
+    private static TestRequest request() {
+        byte[] payload = new byte[ORIGINAL_PAYLOAD_SIZE];
+        Arrays.fill(payload, (byte) 1);
+        return TestRequest.newBuilder().setName(new String(payload, StandardCharsets.US_ASCII)).build();
+    }
+
+    private void assertSuccessful(final GrpcMessageEncoding encoding) throws ExecutionException, InterruptedException {
+        client.test(new TestMetadata(encoding), request()).toFuture().get();
```

Comment:
Not a big fan of the request-encoding been defined as part of the metadata. I played with a couple of alternatives,
eg. `client.withCompression(...).test(...)` which is more fluent, but it has the problem of having to support multiple classes every time we add more fields, to avoid the API misbehavior of `client.withCompression(...).withCompression(..)...` 

If we don't mind this API issue (which personally I have mixed feelings), then we could use that notation, lose the field from the metadata, and allow for even more flexible use-cases like
- `.withCompression(request_encoding, response_encodings)` which can override the response encodings per call, rather than use the default ones (defined per client).

On the other hand, we can always start this way, and re-visit it in the future once we collect some feedback from end-users.